### ISSUE TITLE
Support file-like objects in geopandas.read_file

### ISSUE THREAD

**Issue by jorisvandenbossche** [geopandas/geopandas#1323]
Currently, `read_file` supports file paths and urls. And for urls, we download the data, and put that in a fiona "BytesCollection". 
I *suppose* (but didn't yet look in detail) that something similar can be done for file-like objects as well (in addition to urls). Eg if one already opened a file, or has a bytes object from a response, .. that represents the data.

This would make it more similar to pandas in what types of objects it accepts for the path / source.

**Comment by ian-r-rose** [comment]
Supporting file-like objects would also allow for some nice integrations between [fsspec](https://github.com/intake/filesystem_spec), [intake](https://github.com/intake/intake), and [intake_geopandas](https://github.com/intake/intake_geopandas). In particular, that should make caching certain geospatial data sources much easier, c.f. https://github.com/intake/intake_geopandas/pull/10

**Comment by Sangarshanan** [comment]
It seems like a duplicate of https://github.com/geopandas/geopandas/issues/388 . I would like to take a crack at this looking at the abundance of resourceful comments in the issue and the ones in fiona

**Comment by Sangarshanan** [comment]
Converting StringIO objects and text streams to a string can help read them 

Just needed to add this line 
```
if (type(filename) == TextIOWrapper) or (type(filename) == StringIO):
    path_or_bytes = filename.read()
    reader = fiona.open
```
Pathlib paths, binary streams and bytesIO already work 
```
## pathlib.Path (supported)
from pathlib import Path
path_object = Path("test.geojson")
gpd.read_file(path_object)

## To read binary stream & bytesIO (supported)
f = open("test.geojson", "rb")
gpd.read_file(f)
```

**Comment by martinfleis** [comment]
Closed by #1329
